### ISSUE TITLE
Hook up node-gyp to CI

### ISSUE THREAD

**Issue by bnoordhuis** [nodejs/build#206]
Can we add https://github.com/nodejs/node-gyp to the CI?  There isn't much of a test suite currently but I'm hoping that will change over time.  `npm test` writes tap output, FWIW.

**Comment by orangemocha** [comment]
Sounds like a good idea. Are you looking for something like node-test-pull-request / node-test-commit to test changes in this repo before they are merged?

**Comment by bnoordhuis** [comment]
That would be ideal, yes.

**Comment by orangemocha** [comment]
I'll add it to my queue. If someone else wants to pick it up, be welcome :wink: 

I am assuming it would make sense to run on all the core support platforms (see https://github.com/nodejs/build#test-configurations). 

This new CI job would re-use the current Jenkins slaves. In order to avoid port conflicts / interfering node instances, it would have to run exclusively from other node-test-* jobs, so it will essentially be competing for those slaves. I am assuming that it won't be an issue, as we don't get too many pull requests on this repo.

**Comment by bnoordhuis** [comment]
I don't think so either but FWIW, the tests don't start any servers and I don't think they ever will, so it's probably fine to run them in parallel with other jobs.

**Comment by bnoordhuis** [comment]
Friendly ping.  Has there been any progress on this?  There are pull requests I'd like to land that are held up by this.

**Comment by jbergstroem** [comment]
How do you see this work? Build node similar to how we do in test-pr, pull node-gyp and test?

**Comment by bnoordhuis** [comment]
I don't think there's a need to build node but it could be valuable to test node-gyp against a range of node release binaries on different platforms.  The current test suite mostly consists of self-tests, it doesn't try to build any add-ons (yet - that's probably the next step.)

**Comment by orangemocha** [comment]
I haven't made any progress, sorry. Let's bump it up.

I see a few ways to go about this:

1. A job that can test any ref on the node-gyp repo. Instead of including the logic for building node, it downloads node from https://nodejs.org/dist/ with a version parameter. This can be run on a matrix of a handful of platforms. This is similar to what we did for https://ci.nodejs.org/job/serialport-test-commit-windows/build?delay=0sec, though that job is only limited to Windows. (Implementation note: perhaps we can reuse the rebase job and the temp repo approach).
2. Extend node-test-commit to take an optional ref from the node-gyp repo. In the beginning of the job, pull the ref and overwrite what's in the node tree. Then proceed to build node as usual. Add node-gyp tests to node-test-commit. This probably adds unnecessary complexity to the jobs.
3. Start directing node-gyp changes to the node repo directly. I think we have already accepted the fact that the only supported version of node-gyp is the one that ships with node, so it does make sense to test the node/node-gyp combo from the same branch. However this introduces additional PR noise in the node repo.
4. Similar to 3, but with submodules/subrepos. I wish we had a better way to manage our dependencies with a sub-module like approach. But I am just rambling now...

So I think option 1. is the most practical at the moment. @joaocgreis is there a better way?

**Comment by orangemocha** [comment]
This was discussed at the last WG meeting. The plan is to do something similar to  https://ci.nodejs.org/job/serialport-test-commit-windows/build?delay=0sec but to have it run on multiple versions of node rather than a specific version. The node versions and platforms would be part of the job's matrix.

A similar issue was raised about NAN, even though I can't find a specific issue about that. We agreed at the WG meeting that the same job structure would apply to NAN.

And once we have figured out how to run node-gyp and NAN tests in CI, we might want to add (part of) those tests to node-test-commit, to prevent breakage due to changes in node.

This is all still on my and @joaocgreis' queue. Of course if anyone wants to make progress on it sooner we won't complain :wink:

**Comment by bnoordhuis** [comment]
@joaocgreis @orangemocha Have you had time to look into this?

**Comment by rvagg** [comment]
do we really have enough tests to make this worthwhile?

**Comment by bnoordhuis** [comment]
Yes.  It's steadily growing.

**Comment by joaocgreis** [comment]
Finally got the ball rolling on this one, sorry for the delay.

@bnoordhuis can you take a look and try https://ci.nodejs.org/view/All/job/nodegyp-test-commit/ ? Only two platforms now, but the job should do the right thing. What would be the next priority?

- Add more platforms, all as configurations in the same job? (much easier to check results IMO)
- Add more platforms with subjobs, as `node-test-commit`? (allows launching only one subjob)
- Add tap parser? (Is there a way to make `npm test` produce a `test.tap` file?)
- Something else?

**Comment by bnoordhuis** [comment]
@joaocgreis Thanks, looks great.

> Add more platforms, all as configurations in the same job?

Sounds good.  I don't expect sub-jobs is something we'll need.

> Add tap parser? (Is there a way to make npm test produce a test.tap file?)

tape produces tap by default but npm seems to swallow the output.  If I run `node node_modules/.bin/tape test/test-*`, it works.

**Comment by joaocgreis** [comment]
Added https://ci.nodejs.org/job/nodegyp-test-pull-request/ to spawn nodegyp-test-commit with multiple versions of node, but for now the versions still have to be manually updated in the job config and there is still no rebase support.

**Comment by maclover7** [comment]
Looks like the job has been created, and is working well so far. Please open up a new issue if additional work is needed (more platforms, etc.)

**Comment by refack** [comment]
Ref: https://github.com/nodejs/build/issues/1294
